compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -917,6 +943,161 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
+  SmallVector<OpFoldResult> tileOffsetsBasis = getAsIndexOpFoldResult(
+      ctx, filterSwizzledShape(swizzle, [](TileSwizzle::Dim d) {
+        return d.kind == TileSwizzle::Dim::Kind::CrossThread;
+      }));
+  SmallVector<OpFoldResult> tileOffsets =
+      builder
+          .create<affine::AffineDelinearizeIndexOp>(loc, threadId,
+                                                    tileOffsetsBasis)
+          ->getResults();
+
+  // Strides are trivial: each slice is contiguous along the *expanded* dims
+  // even if it may not be contiguous in the flattened layout.
+  SmallVector<OpFoldResult> tileStrides(tileSizes.size(),
+                                        builder.getIndexAttr(1));
+
+  offsets.append(tileOffsets);
+  sizes.append(tileSizes);
+  strides.append(tileStrides);
+
+  return success();
+}
+
+static void incrementIndices(MutableArrayRef<int64_t> indices,
+                             ArrayRef<int64_t> sizes) {
+  int rank = indices.size();
+  for (int i = rank - 1; i >= 0; --i) {
+    ++indices[i];
+    if (indices[i] == sizes[i]) {
+      indices[i] = 0;
+    } else {
+      break;
+    }
+  }
+}
+
+static Value flattenVector(OpBuilder &builder, Location loc, Value value) {
+  Type type = value.getType();
+  VectorType vectorType = llvm::dyn_cast<VectorType>(type);
+  assert(vectorType);
+  if (vectorType.getRank() <= 1) {
+    return value;
+  }
+  VectorType flatVectorType =
+      VectorType::get(SmallVector<int64_t>{vectorType.getNumElements()},
+                      vectorType.getElementType());
+  return builder.create<vector::ShapeCastOp>(loc, flatVectorType, value);
+}
+
+static SmallVector<Value>
+distributeMmaFragmentToIntrinsics(OpBuilder &builder, Location loc, Value value,
+                                  const TileSwizzle &swizzle) {
+  auto internalShape = filterSwizzledShape(swizzle, [](TileSwizzle::Dim dim) {
+    return dim.kind == TileSwizzle::Dim::Kind::Internal;
+  });
+  auto crossIntrinsicShape =
+      filterSwizzledShape(swizzle, [](TileSwizzle::Dim dim) {
+        return dim.kind == TileSwizzle::Dim::Kind::CrossIntrinsic;
+      });
+  int rank = internalShape.size();
+  auto strides = SmallVector<int64_t>(rank, 1);
+  int64_t crossIntrinsicCount =
+      std::reduce(crossIntrinsicShape.begin(), crossIntrinsicShape.end(), 1,
+                  std::multiplies<int64_t>());
+  SmallVector<Value> distributedValues(crossIntrinsicCount);
+  SmallVector<int64_t> indices(rank, 0);
+  for (Value &distributedValue : distributedValues) {
+    Value extract = builder.create<vector::ExtractStridedSliceOp>(
+        loc, value, indices, internalShape, strides);
+    distributedValue = flattenVector(builder, loc, extract);
+    incrementIndices(indices, crossIntrinsicShape);
+  }
+  return distributedValues;
+}
+
+// Generates amdgpu.mfma/wmma operation on the given inputs for this attribute
+// type.
+FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
```

Comment:
WMMA is not supported, so should probably fail in that case and indicate enabling it in the doc as TODO. Same with other functions here.

Reply:
We are already asserting in `getIntrinsicSwizzle`, at the root cause.  I will just add TODO comments.